File: Leetcode/3537 - Fill a Special Grid/solution.py
class Solution:
    def specialGrid(self, n: int) -> List[List[int]]:
        max_dimension = 1 << n
        grid = [[-1 for _ in range(max_dimension)] for _ in range(max_dimension)]

        counter, head = (0, 0)
        queue = [{ "x": 0, "y": max_dimension - 1, "max_step": max_dimension }]

        while (head < len(queue)):
            x = queue[head]['x']
            y = queue[head]['y']
            max_step = queue[head]['max_step']
            head += 1
            xi = x
            yi = y

            if (max_step == 1):
                grid[xi][yi] = counter
                counter += 1
            else:
                half = (max_step // 2)
                queue.append({ "x": x, "y": y, "max_step": half })  # Top-Right
                x += half
                queue.append({ "x": x, "y": y, "max_step": half })  # Bottom-Right
                y -= half
                queue.append({ "x": x, "y": y, "max_step": half })  # Bottom-Left
                x -= half
                queue.append({ "x": x, "y": y, "max_step": half })  # Top-Left

        return grid 

File: Leetcode/3537 - Fill a Special Grid/solution.ts
function specialGrid(n: number): number[][] {
    const maxDimension: number = 1 << n;
    const grid: number[][] = Array.from({ length: maxDimension }, () =>
        Array(maxDimension).fill(-1)
    );

    let counter: number = 0;
    const queue: { x: number; y: number; maxStep: number }[] = [
        { x: 0, y: maxDimension - 1, maxStep: maxDimension }
    ];
    let head: number = 0;

    while (head < queue.length) {
        let { x, y, maxStep } = queue[head++];
        const xi: number = Math.floor(x);
        const yi: number = Math.floor(y);

        if (maxStep === 1) {
            grid[xi][yi] = counter++;
        } else {
            const half = maxStep / 2;
            queue.push({ x: x, y: y, maxStep: half });  // Top-Right
            x += half;
            queue.push({ x: x, y: y, maxStep: half });  // Bottom-Right
            y -= half;
            queue.push({ x: x, y: y, maxStep: half });  // Bottom-Left
            x -= half;
            queue.push({ x: x, y: y, maxStep: half });  // Top-Left
        }
    }

    return grid;
}